// @nteract/presentational-components/lib/components/input.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Input = void 0;

var _style = _interopRequireDefault(require("styled-jsx/style"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var React = _interopRequireWildcard(require("react"));

var Input =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2.default)(Input, _React$Component);

  function Input() {
    (0, _classCallCheck2.default)(this, Input);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Input).apply(this, arguments));
  }

  (0, _createClass2.default)(Input, [{
    key: "render",
    value: function render() {
      if (this.props.hidden) {
        return null;
      }

      return React.createElement("div", {
        className: "jsx-1290819329" + " " + "input-container"
      }, this.props.children, React.createElement(_style.default, {
        styleId: "1290819329",
        css: ".input-container.jsx-1290819329{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-flex-direction:row;-ms-flex-direction:row;flex-direction:row;}.input-container.invisible.jsx-1290819329{height:34px;}.input-container.jsx-1290819329 .prompt{-webkit-flex:0 0 auto;-ms-flex:0 0 auto;flex:0 0 auto;}.input-container.jsx-1290819329 .input{-webkit-flex:1 1 auto;-ms-flex:1 1 auto;flex:1 1 auto;overflow:auto;background-color:var(--theme-cell-input-bg,#fafafa);}"
      }));
    }
  }]);
  return Input;
}(React.Component);

exports.Input = Input;
(0, _defineProperty2.default)(Input, "defaultProps", {
  children: null,
  hidden: false
});
