// @nteract/directory-listing/lib/components/entry.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Entry = void 0;

var _style = _interopRequireDefault(require("styled-jsx/style"));

var _typeof2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/typeof"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var React = _interopRequireWildcard(require("react"));

var _reactHotLoader = require("react-hot-loader");

var _icon = require("./icon");

var _name = require("./name");

var _lastsaved = require("./lastsaved");

(function () {
  var enterModule = require('react-hot-loader').enterModule;

  enterModule && enterModule(module);
})();

var Entry =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2.default)(Entry, _React$Component);

  function Entry() {
    (0, _classCallCheck2.default)(this, Entry);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Entry).apply(this, arguments));
  }

  (0, _createClass2.default)(Entry, [{
    key: "render",
    value: function render() {
      return React.createElement("tr", {
        className: "jsx-3511019751" + " " + "directory-entry"
      }, React.Children.map(this.props.children, function (child) {
        if ((0, _reactHotLoader.areComponentsEqual)(child.type, _icon.Icon) || (0, _reactHotLoader.areComponentsEqual)(child.type, _name.Name) || (0, _reactHotLoader.areComponentsEqual)(child.type, _lastsaved.LastSaved)) {
          return React.cloneElement(child, {
            className: (0, _typeof2.default)(child.props.className) === String && child.props.className !== "" ? child.props.className + " directory-entry-field" : "directory-entry-field"
          });
        } else {
          return React.createElement("td", {
            className: "jsx-3511019751" + " " + "directory-entry-field"
          }, child);
        }
      }), React.createElement(_style.default, {
        styleId: "3511019751",
        css: "tr.jsx-3511019751{border-top:1px solid #eaecef;}tr.jsx-3511019751:hover{background-color:#f6f8fa;-webkit-transition:background-color 0.1s ease-out;transition:background-color 0.1s ease-out;}tr.jsx-3511019751:first-child{border-top:none;}tr.jsx-3511019751:last-child{border-bottom:none;}"
      }));
    }
  }, {
    key: "__reactstandin__regenerateByEval",
    // @ts-ignore
    value: function __reactstandin__regenerateByEval(key, code) {
      // @ts-ignore
      this[key] = eval(code);
    }
  }]);
  return Entry;
}(React.Component);

exports.Entry = Entry;
(0, _defineProperty2.default)(Entry, "defaultProps", {
  children: null
});
;

(function () {
  var reactHotLoader = require('react-hot-loader').default;

  var leaveModule = require('react-hot-loader').leaveModule;

  if (!reactHotLoader) {
    return;
  }

  reactHotLoader.register(Entry, "Entry", "/Users/kyle/code/src/github.com/nteract/nteract/packages/directory-listing/src/components/entry.js");
  leaveModule(module);
})();

;
