// @nteract/directory-listing/lib/components/lastsaved.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LastSaved = void 0;

var _style = _interopRequireDefault(require("styled-jsx/style"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var React = _interopRequireWildcard(require("react"));

var _timeago = _interopRequireDefault(require("@nteract/timeago"));

(function () {
  var enterModule = require('react-hot-loader').enterModule;

  enterModule && enterModule(module);
})();

var LastSaved =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2.default)(LastSaved, _React$Component);

  function LastSaved() {
    (0, _classCallCheck2.default)(this, LastSaved);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(LastSaved).apply(this, arguments));
  }

  (0, _createClass2.default)(LastSaved, [{
    key: "render",
    value: function render() {
      return React.createElement("td", {
        className: "jsx-835568519" + " " + "timeago"
      }, this.props.lastModified != null && React.createElement(_timeago.default, {
        date: this.props.lastModified
      }), React.createElement(_style.default, {
        styleId: "835568519",
        css: ".timeago.jsx-835568519{text-align:right;color:#6a737d;padding-right:10px;}"
      }));
    }
  }, {
    key: "__reactstandin__regenerateByEval",
    // @ts-ignore
    value: function __reactstandin__regenerateByEval(key, code) {
      // @ts-ignore
      this[key] = eval(code);
    }
  }]);
  return LastSaved;
}(React.Component);

exports.LastSaved = LastSaved;
(0, _defineProperty2.default)(LastSaved, "defaultProps", {
  lastModified: null
});
;

(function () {
  var reactHotLoader = require('react-hot-loader').default;

  var leaveModule = require('react-hot-loader').leaveModule;

  if (!reactHotLoader) {
    return;
  }

  reactHotLoader.register(LastSaved, "LastSaved", "/Users/kyle/code/src/github.com/nteract/nteract/packages/directory-listing/src/components/lastsaved.js");
  leaveModule(module);
})();

;
