// @nteract/directory-listing/lib/components/listing.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Listing = void 0;

var _style = _interopRequireDefault(require("styled-jsx/style"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var React = _interopRequireWildcard(require("react"));

(function () {
  var enterModule = require('react-hot-loader').enterModule;

  enterModule && enterModule(module);
})();

var Listing =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2.default)(Listing, _React$Component);

  function Listing() {
    (0, _classCallCheck2.default)(this, Listing);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Listing).apply(this, arguments));
  }

  (0, _createClass2.default)(Listing, [{
    key: "render",
    value: function render() {
      return React.createElement(React.Fragment, null, React.createElement("div", {
        className: "jsx-1481995636" + " " + "listing-root"
      }, React.createElement("table", {
        className: "jsx-1481995636"
      }, React.createElement("tbody", {
        className: "jsx-1481995636"
      }, this.props.children))), React.createElement(_style.default, {
        styleId: "1481995636",
        css: "table.jsx-1481995636{width:100%;border-collapse:collapse;border-radius:2px;border-spacing:0;}.listing-root.jsx-1481995636{padding:0px 0px 20px 0px;}"
      }));
    }
  }, {
    key: "__reactstandin__regenerateByEval",
    // @ts-ignore
    value: function __reactstandin__regenerateByEval(key, code) {
      // @ts-ignore
      this[key] = eval(code);
    }
  }]);
  return Listing;
}(React.Component);

exports.Listing = Listing;
(0, _defineProperty2.default)(Listing, "defaultProps", {
  children: null
});
;

(function () {
  var reactHotLoader = require('react-hot-loader').default;

  var leaveModule = require('react-hot-loader').leaveModule;

  if (!reactHotLoader) {
    return;
  }

  reactHotLoader.register(Listing, "Listing", "/Users/kyle/code/src/github.com/nteract/nteract/packages/directory-listing/src/components/listing.js");
  leaveModule(module);
})();

;
