// @nteract/directory-listing/lib/components/name.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Name = void 0;

var _style = _interopRequireDefault(require("styled-jsx/style"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var React = _interopRequireWildcard(require("react"));

(function () {
  var enterModule = require('react-hot-loader').enterModule;

  enterModule && enterModule(module);
})();

var Name =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2.default)(Name, _React$Component);

  function Name() {
    (0, _classCallCheck2.default)(this, Name);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Name).apply(this, arguments));
  }

  (0, _createClass2.default)(Name, [{
    key: "render",
    value: function render() {
      return React.createElement("td", {
        className: "jsx-534759121" + " " + "name"
      }, this.props.children, React.createElement(_style.default, {
        styleId: "534759121",
        css: ".name.jsx-534759121{vertical-align:middle;font-size:0.9em;padding:8px;}.name.jsx-534759121 a{-webkit-text-decoration:none;text-decoration:none;}.name.jsx-534759121 a:hover{-webkit-text-decoration:underline;text-decoration:underline;outline-width:0;}"
      }));
    }
  }, {
    key: "__reactstandin__regenerateByEval",
    // @ts-ignore
    value: function __reactstandin__regenerateByEval(key, code) {
      // @ts-ignore
      this[key] = eval(code);
    }
  }]);
  return Name;
}(React.Component);

exports.Name = Name;
(0, _defineProperty2.default)(Name, "defaultProps", {
  children: null
});
;

(function () {
  var reactHotLoader = require('react-hot-loader').default;

  var leaveModule = require('react-hot-loader').leaveModule;

  if (!reactHotLoader) {
    return;
  }

  reactHotLoader.register(Name, "Name", "/Users/kyle/code/src/github.com/nteract/nteract/packages/directory-listing/src/components/name.js");
  leaveModule(module);
})();

;
