// @nteract/mathjax/lib/context.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/assertThisInitialized"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var React = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _loadScript = _interopRequireDefault(require("./load-script"));

//  strict

/* global MathJax */

/**
 * Context for loading MathJax
 */
var Context =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2.default)(Context, _React$Component);

  function Context(props) {
    var _this;

    (0, _classCallCheck2.default)(this, Context);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Context).call(this, props));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)((0, _assertThisInitialized2.default)(_this)), "onLoad", function () {
      if (!MathJax || !MathJax.Hub) {
        _this.props.onError(new Error("MathJax not really loaded even though onLoad called"));

        return;
      }

      var options = _this.props.options;
      MathJax.Hub.Config(options);
      MathJax.Hub.Register.StartupHook("End", function () {
        if (!MathJax) {
          _this.props.onError(new Error("MathJax became undefined in the middle of processing"));

          return;
        }

        MathJax.Hub.processSectionDelay = _this.props.delay;

        if (_this.props.didFinishTypeset) {
          _this.props.didFinishTypeset();
        }
      });
      MathJax.Hub.Register.MessageHook("Math Processing Error", function (message) {
        if (_this.props.onError) {
          _this.props.onError(new Error(message));
        }
      });

      if (_this.props.onLoad) {
        _this.props.onLoad();
      }

      _this.setState({
        loaded: true
      });
    });
    _this.state = {
      loaded: false
    };
    return _this;
  }

  (0, _createClass2.default)(Context, [{
    key: "getChildContext",
    value: function getChildContext() {
      return {
        // Here we see if MathJax is defined globally by running a typeof on a
        // potentially not set value then explicitly setting the MathJax context
        // to undefined.
        MathJax: typeof MathJax === "undefined" ? undefined : MathJax,
        input: this.props.input,
        MathJaxContext: true
      };
    }
  }, {
    key: "componentDidMount",
    value: function componentDidMount() {
      var src = this.props.src;

      if (src == null) {
        return this.onLoad();
      }

      if (!this.props.loader) {
        (0, _loadScript.default)(src, this.onLoad);
      } else {
        this.props.loader(this.onLoad);
      }
    }
  }, {
    key: "render",
    value: function render() {
      if (!this.state.loaded && !this.props.noGate) {
        return this.props.loading;
      }

      return this.props.children;
    }
  }]);
  return Context;
}(React.Component);

(0, _defineProperty2.default)(Context, "defaultProps", {
  src: "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML",
  input: "tex",
  didFinishTypeset: null,
  loader: null,
  delay: 0,
  options: {},
  loading: null,
  noGate: false,
  onLoad: null,
  onError: function onError(err) {
    console.error(err);
  }
});
Context.childContextTypes = {
  MathJax: _propTypes.default.object,
  input: _propTypes.default.string,
  MathJaxContext: _propTypes.default.bool
};
var _default = Context;
exports.default = _default;
