// @nteract/presentational-components/src/syntax-highlighter/theme.js
// @flow

export default {
  hljs: {
    display: "block",
    overflowX: "auto",
    padding: "0.5em",
    background: "var(--cm-background, #fafafa)",
    color: "var(--cm-color, black)",
    fontFamily: '"Source Code Pro", monospace',
    // Conform to our CodeMirror-line formatting
    fontSize: "14px",
    lineHeight: "1.21429em"
  },
  "hljs-name": {
    fontWeight: 600
  },
  "hljs-strong": {
    fontWeight: 600
  },
  "hljs-code": {
    fontStyle: "italic",
    color: "#888"
  },
  "hljs-emphasis": {
    fontStyle: "italic"
  },
  "hljs-tag": {
    color: "#170"
  },
  "hljs-variable": {
    color: "var(--cm-variable, black)"
  },
  "hljs-template-variable": {
    color: "var(--cm-variable, black)"
  },
  "hljs-selector-id": {
    color: "#ade5fc"
  },
  "hljs-selector-class": {
    color: "#ade5fc"
  },
  "hljs-string": {
    color: "var(--cm-string, #a22)"
  },
  "hljs-bullet": {
    color: "#d36363"
  },
  "hljs-type": {
    color: "#085"
  },
  "hljs-title": {
    color: "#05a"
  },
  "hljs-section": {
    color: "#05a"
  },
  "hljs-attribute": {
    color: "#00c"
  },
  "hljs-quote": {
    color: "#090"
  },
  "hljs-built_in": {
    color: "var(--cm-builtin, #077)"
  },
  "hljs-builtin-name": {
    color: "var(--cm-builtin, #077)"
  },
  "hljs-number": {
    color: "var(--cm-number, #3a3)"
  },
  "hljs-symbol": {
    color: "#d36363"
  },
  "hljs-keyword": {
    color: "var(--cm-keyword, blue)",
    fontWeight: 700
  },
  "hljs-selector-tag": {
    color: "#fcc28c"
  },
  "hljs-literal": {
    color: "#fcc28c"
  },
  "hljs-comment": {
    color: "var(--cm-comment, #a86)"
  },
  "hljs-deletion": {
    color: "#333",
    backgroundColor: "#fc9b9b"
  },
  "hljs-regexp": {
    color: "#c6b4f0"
  },
  "hljs-link": {
    color: "var(--cm-link, #3a3)"
  },
  "hljs-meta": {
    color: "var(--cm-meta, #555)"
  },
  "hljs-addition": {
    backgroundColor: "#a2fca2",
    color: "#333"
  }
};
